File: code/client/core/source/operation-graph/CommandInfo.cpp
﻿// <copyright file="CommandInfo.cpp" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

module;

#include <chrono>

export module Soup.Core:CommandInfo;

import Opal;

using namespace Opal;
using namespace std::chrono_literals;

namespace Soup::Core
{
	/// <summary>
	/// The core command information that describes exactly what to run when evaluating an operation.
	/// </summary>
	export class CommandInfo
	{
	public:
		Path WorkingDirectory;
		Path Executable;
		std::vector<std::string> Arguments;

	public:
		CommandInfo() :
			WorkingDirectory(),
			Executable(),
			Arguments()
		{
		}

		CommandInfo(
			Path workingDirectory,
			Path executable,
			std::vector<std::string> arguments) :
			WorkingDirectory(std::move(workingDirectory)),
			Executable(std::move(executable)),
			Arguments(std::move(arguments))
		{
		}

		bool operator ==(const CommandInfo& rhs) const
		{
			return WorkingDirectory == rhs.WorkingDirectory &&
				Executable == rhs.Executable &&
				Arguments == rhs.Arguments;
		}
	};
}

namespace std
{
	template<> struct hash<Soup::Core::CommandInfo>
	{
		std::size_t operator()(Soup::Core::CommandInfo const& value) const noexcept
		{
			std::size_t hashWorkingDirectory = std::hash<std::string>{}(value.WorkingDirectory.ToString());
			std::size_t hashExecutable = std::hash<std::string>{}(value.Executable.ToString());
			std::size_t hashArguments = value.Arguments.size();
			return hashWorkingDirectory ^ (hashExecutable << 1) ^ (hashArguments << 2);
		}
	};
}

File: code/generate-sharp/utilities/operation-graph/command-info.cs
// <copyright file="command-info.cs" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

using System;
using System.Collections.Generic;
using Path = Opal.Path;

namespace Soup.Build.Utilities;

public class CommandInfo : IEquatable<CommandInfo>
{
	/// <summary>
	/// Initializes a new instance of the <see cref="CommandInfo"/> class.
	/// </summary>
	public CommandInfo() :
		this(new Path(), new Path(), [])
	{
	}

	/// <summary>
	/// Initializes a new instance of the <see cref="CommandInfo"/> class.
	/// </summary>
	public CommandInfo(
		Path workingDirectory,
		Path executable,
		IReadOnlyList<string> arguments)
	{
		this.WorkingDirectory = workingDirectory;
		this.Executable = executable;
		this.Arguments = arguments;
	}

	/// <summary>
	/// Equality operator
	/// </summary>
	public bool Equals(CommandInfo? other)
	{
		if (other is null)
			return false;
		return this.WorkingDirectory == other.WorkingDirectory &&
			this.Executable == other.Executable &&
			this.Arguments == other.Arguments;
	}

	public override bool Equals(object? obj)
	{
		return Equals(obj as CommandInfo);
	}

	public override int GetHashCode()
	{
		return (this.WorkingDirectory.GetHashCode() * 0x100000) + (this.Executable.GetHashCode() * 0x1000) + this.Arguments.GetHashCode();
	}

	public static bool operator ==(CommandInfo lhs, CommandInfo other)
	{
		if (lhs is null)
			return other is null;
		return lhs.Equals(other);
	}

	public static bool operator !=(CommandInfo lhs, CommandInfo other)
	{
		return !(lhs == other);
	}

	public Path WorkingDirectory { get; init; }
	public Path Executable { get; init; }
	public IReadOnlyList<string> Arguments { get; init; }
}